client/src/pages/Map/Map.js

```diff
@@ -2,6 +2,7 @@ import React, { useState, useEffect, useRef } from 'react';
 import mapboxgl from 'mapbox-gl';
 import { Alert, Box, Typography } from '@mui/material';
 import { useQueryClient } from 'react-query';
+import { getData } from '@/api/apiUtils';
```

Comment:
Yup, that's exactly the main point of contention for this PR! I know why we use `react-query` - it provides a lot of useful reusable logic related to making queries such as error handling and caching, which is great. I think we should keep using it rather than moving on to something else.

However, using `react-query` takes away control from us in this specific situation. I noticed that there's a bunch of `useEffect`s in `MapLayout` even though using `react-query` is supposed to _reduce_ the need to use the `useEffect` hook.  That's because we make a tree query as soon as `MapLayout` mounts. Now, we must wait for both 1) `react-query` to wait for a response and update the state of `MapLayout` and 2) `Map` to kick off the `mapboxMap.load` callback which also updates the state of `MapLayout`. Now, we need a `useEffect` to kick off when we have both the tree query response and map load, which was buggy in the previous implementation, and is hard to implement in this iteration in a way where it'll be free from bugs in the future.

Before, we combined two updates 1) Kick off a map update side effect when the page loads with a certain tree id and 2) Kick off a map update side effect when a tree id is changed. It kind of worked before when we conflated both. Now, we want to separate map load and tree selection behaviors.

Maybe I don't quite know what you had in mind for "bypassing api/queries and using react-query". What I need is a piece of asynchronous code related to making tree queries that I can `await` in the `mapboxMap.load` callback. `react-query` does not provide that since it abstracts asynchronous code into _React lifecycle state_. I'm open to any alternate suggestions!

Here are some example alternative approaches I tried with `react-query`  to get it to just fire **once** upon page load that doesn't quite work:
* Only call the `useEffect` hook once upon mounting `MapLayout`. This doesn't work since we need to wait for the response of the tree query
* Have the `useEffect` hook be called every time `map` and `currentTreeDb` is updated, but keep some kind of flag that signifies the "first" time and only call it if it's the "first" time and the map is loaded. Doesn't work because map doesn't necessarily load with a tree id in the URL.
* Use the `react-query` `onSuccess` callback. This is called every time the tree query is fetched, including refetches.

Instead, with these changes, we avoid calling `react-query` in this instance here. We then update the `react-query` cache manually to avoid any issues with double calls. We depart from the `react-query` standard in this one instance. However, I would argue the code is much easier to read and reason about. There is a strong reason for why we should do it this way.

Reply:
Yup, I tried a lazy query. The problem is that it calls `onSettled` every time there's a refetch, which is a core feature of `react-query`, and also it runs every single time `currentTreeId` changes.

Addressing each resource:
* Important defaults - yup, played around with this, which spawned https://github.com/waterthetrees/wtt_front/pull/413. 
* `useQuery` API - yup, I experimented with various status state and callbacks
* Lazy queries - In addition to what has already been said, this would require some sort of lock, and I could only think of v complicated approaches

> One thing I noticed is this doesn't handle the vector data /vector tree sharing of hash params(our current prod version doesn't either).
From what was demonstrated at the demo, it does handle vector data. There was a bug at the time you probably reviewed where a sidebar did not appear if no tree data existed for the specific tree, but I pushed a commit to fix that

Reply:
Okay, here's something we can do in pseudocode:
```
If (tree ID in the URL) {
  isEnabled = true;
}
onSettled: (data, error) => {
  if (!error && map.loaded && isEnabled) {
    flyToTree([data.lng, data.lat]);
  }  else if (!error && isEnabled) {
    setQueuedUpFlyTo([data.lng, data.lat]) // If the map takes longer to load than the tree query
  }
  isEnabled = false;
}

in Map.js
map.load(() => {
  if (queuedUpToFlyTo) {
   flyToTree(queuedUpToFlyTo);
   isEnabled = false;
  }
});
```
^ I think even this has a race condition bug in it where `onSettled` can be called right after `map.load` and before the state change rerenders the components. I also don't like how the logic is spread out between multiple components, and I think it'll be even harder to understand when we add the geolocation check.  

I understand the reluctance to diverge from using one pattern everywhere, but I do think this is a case that `react-query` is ill-suited for until a cleaner solution is found. We shouldn't be using `react-query` for use cases where it makes things less understandable.

My argument is:
My solution in the PR is cleaner than the useEffect() that listened for changes to map, currentTreeDb, and currentTreeDataVector state changes implementation. It's encapsulated in a helper function that is called within `map.load` that is actually much easier to update. I will have to add some extra logic here when we add flying to geolocation upon load, and I think future maintainability won't be bad.

`react-query` abstracts asynchronous calls into react lifecycles and that's counterproductive when we want to only make a call as a side effect of an asynchronous call. The lazy queries are good when we want to make a call as a side effect of another `react-query` call or react state change.

Reply:
Ooh, I thought of a much better way of doing it based on your suggestions. Give me time to make a commit with that approach since I'll need to take some time to think about how it'll fit in with other future onLoad side effects we want to introduce - not sure I like this approach yet but it'll keep the use of `react-query`

Reply:
Yeah, sorry. Planning on getting an iteration out today. This is taking a ton of thinking lol in terms of what is the best way for future maintainability

Reply:
@zoobot 
Okay, here's the alternate solution with `react-query`. My mind block was separating updates from subsequent tree clicks from the initial tree query. In this solution, we have two separate tree queries - one in `Map` and one in `MapLayout`. I still don't love this solution because I think could it be slightly easier to introduce a bug - i.e., this useEffect is called every time the component updates unlike the solution in this PR which is guaranteed to only fire once on map load. 

But I admit that it's concise and keeps to the "use react-query everywhere" standard. I'm okay with going with either. Your call @zoobot!

https://github.com/waterthetrees/wtt_front/pull/422/files#diff-9d7e66879ba8b56e3e2be3dfd37cb4655ef5cd17188ca9b82d0301381514d6fe

Reply:
If we go with the other solution, let me merge the commit into this PR instead of merging the other draft PR so we can retain all the comments